ntp-client/src/Network/NTP/Client.hs

```diff
@@ -71,248 +68,308 @@ data NtpStatus =
+       )
+
+data QueryOutcome = Timeout | Result NtpOffset
+
+runQueryLoop ::
+       Tracer IO NtpTrace
+    -> NtpClientSettings
+    -> TVar NtpStatus
+    -> TVar (Maybe [ReceivedPacket])
+    -> These (Socket, [AddrInfo]) (Socket, [AddrInfo])
+    -> IO (Either IOError ())
+runQueryLoop tracer ntpSettings ntpStatus inQueue servers = tryIOError $ forever $ do
+    traceWith tracer NtpTraceClientStartQuery
+    void $ atomically $ writeTVar inQueue $ Just []
+    (_id, outcome) <- runThreads
+    case outcome of
+        Timeout -> do
+            traceWith tracer NtpTraceUpdateStatusQueryFailed
+            atomically $ writeTVar ntpStatus NtpSyncUnavailable
+        Result offset -> do
+             traceWith tracer $ NtpTraceUpdateStatusClockOffset $ getNtpOffset offset
+             atomically $ writeTVar ntpStatus $ NtpDrift offset
+
+    void $ atomically $ writeTVar inQueue Nothing
+    traceWith tracer NtpTraceClientSleeping
+    threadDelayInterruptible ntpStatus $ fromIntegral $ ntpPollDelay ntpSettings
     where
-    -- Receive responses from the network and update NTP client state.
-    loop :: AddrFamily -> Socket.Socket -> IO ()
-    loop addressFamily sock
-        = handle (handleIOException addressFamily) $ forever $ do
-            (bs, _) <- recvFrom sock ntpPacketSize
-            case decodeOrFail $ LBS.fromStrict bs of
-                Left  (_, _, err)    ->
-                    traceWith tracer $ NtpTraceReceiveLoopDecodeError err
-                Right (_, _, packet) ->
-                    handleNtpPacket packet
-
-    -- Restart the @loop@ in case of errors; wait 5s before recreating the
-    -- socket.
-    handleIOException
-        :: AddrFamily
-        -> IOException
+        runThreads
+           = withAsync (sendBoth servers >> loopForever)      $ \sender ->
+             withAsync (timeout      >> return Timeout)   $ \delay ->
+             withAsync (checkReplies >>= return . Result) $ \revc ->
+                    waitAnyCancel [sender, delay, revc]
+
+        timeout = do
+            threadDelay $ fromIntegral $ ntpResponseTimeout ntpSettings
+            traceWith tracer NtpTraceClientWaitingForRepliesTimeout
+
+        checkReplies = atomically $ do
+            q <- readTVar inQueue
+            case q of
+               Nothing -> error "Queue of incomming packets not available"
+               Just l -> case (ntpReportPolicy ntpSettings) l of
+                   Wait -> retry
+                   Report r -> return r
+
+        sendBoth s = case s of
+            This v4 -> send v4
+            That v6 -> send v6
+            These v4 v6 -> send v6 >> send v4
+
+        send (sock, addrs) = forM_ addrs $ \addr -> do
+            p <- mkNtpPacket
+            void $ Socket.ByteString.sendTo sock (LBS.toStrict $ encode p) (setNtpPort $ Socket.addrAddress addr)
+            traceWith tracer NtpTracePacketSent
+            
+        loopForever = forever $ threadDelay 1_000_000
```

Comment:
This loop will wake up every second, it's better to use much longer delay.  An hour (or more) would be ok.

Reply:
How about 500_000_000 ?

Reply:
https://github.com/input-output-hk/ouroboros-network/blob/master/ouroboros-network/src/Ouroboros/Network/Protocol/LocalTxSubmission/Client.hs#L47
and a few other places (note that the `threadDelay` there is not the same as one you're using here, hence the missing `0`s).

Reply:
https://github.com/input-output-hk/ouroboros-network/blob/master/ouroboros-network/src/Ouroboros/Network/Protocol/LocalTxSubmission/Client.hs#L47
43200s is a day (but only the 12h without the night)
I just did not want to cause some Int overrun and not sleep too long.
I can change it to 43_200_000_000 but some docs say only 2^29 is safe for Ints.
(Is Karl running cardano on 32Bit systems ? I can't remember)

Reply:
We can just use `maxBound` then :)